### ISSUE TITLE
[API Proposal]: Add custom polynomial support for Crc32 and Crc64 class

### ISSUE THREAD

**Issue by medo64** [dotnet/runtime#78063]
### Background and motivation

There are many variants of CRC32 algorithm that differ only in polynomials used. The current implementation only supports IEEE 802.3 polynomial externally. It would be ideal if class would allow for defining custom polynomial thus supporting other CRC-32 variants (e.g. ISCSI, MPEG-2, etc.)

Looking at source, there is a `Generate` method in [Crc32ReflectedTable](https://source.dot.net/System.IO.Hashing/R/3e99e17f65099669.html) so basic plumbing for custom polynomials seems to be there already.

As different polynomials use different initial and output XOR values, that would need to be added.

### API Proposal

```csharp
namespace namespace System.IO.Hashing

public sealed partial class Crc32 : NonCryptographicHashAlgorithm
{
    public Crc32(ReadOnlySpan<byte> reflectedPolynomial);
    public Crc32(ReadOnlySpan<byte> reflectedPolynomial,
                 ReadOnlySpan<byte> initialValue,
                 ReadOnlySpan<byte> xorValue,
                 bool reflectIn,
                 bool reflectOut);
}
```

### API Usage

```csharp
var crc = new Crc32(new byte[] { 0x8F, 0x6E, 0x37, 0xA0 });
crc.Append(Encoding.ASCII.GetBytes("123456789"));
var result = crc.GetCurrentHash();
```

**Comment by teo-tsirpanis** [comment]
The polynomial values should be integers, not spans.

**Comment by medo64** [comment]
> The polynomial values should be integers, not spans.

I tend to agree with you but I proposed it to be spans for the following two reasons:

1. I believe the proper type would be uint and unsigned integers used to be frowned upon in the public API

2. Most of functions that output hash (e.g. GetCurrentHash or GetHashAndReset) actually use spans for hash value. One function that doesn't follow that rule is GetCurrentHash that returns bytes. Since existing class/interface already uses spans for hash outputs, it seemed appropriate to accept hash in the same venue.

**Comment by teo-tsirpanis** [comment]
> unsigned integers used to be frowned upon in the public API

Used to be. Not anymore. You might think of CLS compliance, which is a concept no longer relevant.

The disadvantage of accepting spans is that it falsely communicates to the user that they can pass a span of any length, while only spans of four or eight bytes are allowed.

**Comment by normj** [comment]
Amazon S3 recently added new data integrity checks with CRC64 NVME being the default. https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html

As maintainers of the AWS .NET SDK we don't have a good option for supporting this. I believe if this feature request is done to support for custom polynomial we could add this support.

Sorry for double posting. I started a discussion here but seems to have fallen off the radar: https://github.com/dotnet/runtime/discussions/112771

**Comment by stephentoub** [comment]
@bartonjs, opinions on this?

**Comment by bartonjs** [comment]
Conceptually, it seems fine.  I don't know offhand if there are characteristics of a polynomial that should be tested for and rejected if they meet/don't (e.g. the 0 polynomial is probably bad?).

Since it is instantiable it is theoretically something someone might pass around, and I wonder if confusion would happen when you're expecting CRC-32 and get CRC-32C or some other variant... but that would just be "maybe it should be in a new type named `CustomCrc32`, or something", not opposition.  And I'm not sure if it's an actual concern I have, or just a answer to "what could go wrong?"

**Comment by normj** [comment]
Is there anything I can do to bump up the priority for this?

**Comment by filipnavara** [comment]
We already have a need for this internally in the ILC compiler:
https://github.com/dotnet/runtime/blob/12fc2d2ff39e7ccc79a7d76c43cb476f4a77ec4c/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/CoffObjectWriter.cs#L1059-L1129

This would be a good opportunity to clean that up.

**Comment by IDisposable** [comment]
Given that the current framework in [/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs](https://github.com/dotnet/runtime/blob/379af99ab7dacd5daa924f6b7df9907af3bf93d9/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs#L766-L961) exposes Crc32/Crc32C by calling to enabled intrinsic for `Sse42.Crc32C` or `Crc32.ComputeCrc32C` or `Crc32.Arm64.ComputeCrc32C` or to a private `Crc32Fallback` class, **this is not a trivial change**.

CRC-32C (Catagnoli) which is used for  SSE4.1 crc 

Polynomial representations [ref](https://en.wikipedia.org/wiki/Cyclic_redundancy_check#CRC-32_algorithm)

Name | Normal | Reversed | Reciprocal | Reversed reciprocal | SSE4.2 | ARM | 
-- | -- | -- | -- | -- | -- | --
CRC-32  | 0x04C11DB7 | 0xEDB88320 | 0xDB710641 | 0x82608EDB |  | [CRC32B, CRC32H, CRC32W, CRC32X (A64)](https://developer.arm.com/documentation/dui0801/l/A64-General-Instructions/CRC32B--CRC32H--CRC32W--CRC32X--A64-?lang=en)
CRC-32C (Castagnoli) | 0x1EDC6F41 | 0x82F63B78 | 0x05EC76F1 | 0x8F6E37A0 | CRC32 | [CRC32CB, CRC32CH, CRC32CW, CRC32CX (A64)](https://developer.arm.com/documentation/dui0801/l/A64-General-Instructions/CRC32CB--CRC32CH--CRC32CW--CRC32CX--A64-?lang=en)
CRC-32K (Koopman) | 0x741B8CD7 | 0xEB31D82E | 0xD663B05D | 0xBA0DC66B | |

**Comment by bartonjs** [comment]
Late thoughts from the back of my mind: For the API shape, aside from the "do we want a different class here?" I remember from when I first was looking into CRC that there's a .... residue?  I feel like the residue requires knowing if any particular polynomial requires a Big- or Little- Endian output encoding... at least because I think our CRC-32 is LE and our CRC-64 is BE to make them work out.

We could, of course, just try it both ways in the ctor.  But having the ctor check that all the arguments are sound feels... expensive.  For "const generics" it could be done once per process.  Though I guess with ctor parameters it could be once per process too, just needs to use the parameter-set as a lookup into a dictionary.

**Comment by bartonjs** [comment]
Fixed by #124117.
